crates/common/certificate/src/parse_root_certificate/pkcs11.rs

```diff
@@ -0,0 +1,415 @@
+            rustls::Error::General(format!("pkcs11: Failed to sign message: {:?}", err))
+        })?;
+
+        // Split raw signature into r and s values (assuming 32 bytes each)
+        trace!("Signature (raw) len={:?}", signature_raw.len());
+        let r_bytes = &signature_raw[0..32];
+        let s_bytes = &signature_raw[32..];
+        let signature_asn1 = format_asn1_ecdsa_signature(r_bytes, s_bytes).map_err(|err| {
+            rustls::Error::General(format!("pkcs11: Failed to format signature: {:?}", err))
+        })?;
+        trace!(
+            "Encoded ASN.1 Signature: len={:?} {:?}",
+            signature_asn1.len(),
+            signature_asn1
+        );
+        Ok(signature_asn1)
+    }
+
+    fn scheme(&self) -> SignatureScheme {
+        trace!("Using Signature scheme: {:?}", self.scheme.as_str());
+        self.scheme
+    }
+}
+
+#[derive(Debug)]
+pub struct Pkcs11Resolver {
+    pub chain: Vec<CertificateDer<'static>>,
+    pub signing_key: Arc<dyn SigningKey>,
+}
+
+impl ResolvesClientCert for Pkcs11Resolver {
+    fn resolve(
+        &self,
+        _acceptable_issuers: &[&[u8]],
+        _sigschemes: &[SignatureScheme],
+    ) -> Option<Arc<CertifiedKey>> {
+        Some(Arc::new(CertifiedKey {
+            cert: self.chain.clone(),
+            key: self.signing_key.clone(),
+            ocsp: None,
+        }))
+    }
+
+    fn has_certs(&self) -> bool {
+        true
+    }
+}
+
+#[derive(Debug)]
+pub struct ECSigningKey {
+    pkcs11: PKCS11,
+}
+
+impl ECSigningKey {
+    fn supported_schemes(&self) -> &[SignatureScheme] {
+        &[
+            SignatureScheme::ECDSA_NISTP256_SHA256,
+            SignatureScheme::ECDSA_NISTP384_SHA384,
+            SignatureScheme::ECDSA_NISTP521_SHA512,
+        ]
+    }
+}
+
+impl SigningKey for ECSigningKey {
+    fn choose_scheme(&self, offered: &[SignatureScheme]) -> Option<Box<dyn Signer>> {
+        trace!("Offered signature schemes. offered={:?}", offered);
+        let supported = self.supported_schemes();
+        for scheme in offered {
+            if supported.contains(scheme) {
+                trace!("Matching scheme: {:?}", scheme.as_str());
+                return Some(Box::new(PkcsSigner {
+                    pkcs11: self.pkcs11.clone(),
+                    scheme: *scheme,
+                }));
+            }
+        }
+        error!(
+            "Could not find a matching signing scheme. offered={:?}",
+            offered
+        );
+        None
+    }
+
+    fn algorithm(&self) -> SignatureAlgorithm {
+        SignatureAlgorithm::ECDSA
+    }
+}
+
+#[derive(Debug)]
+pub struct RSASigningKey {
+    pkcs11: PKCS11,
+}
+
+impl RSASigningKey {
+    fn supported_schemes(&self) -> &[SignatureScheme] {
+        &[
+            SignatureScheme::RSA_PKCS1_SHA256,
+            SignatureScheme::RSA_PKCS1_SHA384,
+            SignatureScheme::RSA_PKCS1_SHA512,
+            SignatureScheme::RSA_PSS_SHA256,
```

Comment:
Whilst using a Yubikey 5C with a device generated private key, the `SignatureScheme::RSA_PSS_SHA256` signature scheme results in handshake verification errors. This is probably caused by an invalid mapping between the signature scheme and the signing/hashing associated with it (e.g. for direct signing or not).

For now I would recommend removing the `SignatureScheme::RSA_PSS_SHA256` option...the other RSA options above it seem fine.

Reply:
After this scheme was removed, I was able to check that the remaining schemes (`RSA_PKCS1_SHA*`) aren't actually supported by Cumulocity...so at the moment RSA based keys are not supported.

We can live with this for now (as EC seems to be a more modern, (secure?) choice. We can do a follow up task to  understand why the `RSA_PSS_*` schemes aren't working.

Reply:
The [ring docs](https://starry-network.github.io/starry_node/ring/signature/index.html#rsa_pss_-details-rsa-pss-signatures) might offer some clues about this.